pints/_mcmc/_relativistic.py

```diff
@@ -0,0 +1,437 @@
+    moving in dimension ``i`` of negative log posterior space -- which
+    supplements the position (``q_i``) of the particle in parameter space. The
+    particle's motion is dictated by solutions to Hamilton's equations,
+
+    .. math::
+        dq_i/dt &=   \partial H/\partial p_i\\
+        dp_i/dt &= - \partial H/\partial q_i.
+
+    The Hamiltonian is given by,
+
+    .. math::
+        H(q,p) &=       U(q)       +        KE(p)\\
+               &= -\text{log}(p(q|X)p(q)) +
+                    mc^2 (\Sigma_{i=1}^{d} p_i^2 / (mc^2) + 1)^{0.5}
+
+    where ``d`` is the dimensionality of model, ``m`` is the scalar 'mass'
+    given to each particle (chosen to be 1 as default) and ``c`` is the
+    speed of light (chosen to be 10 by default).
+
+    To numerically integrate Hamilton's equations, it is essential to use a
+    sympletic discretisation routine, of which the most typical approach is
+    the leapfrog method,
+
+    .. math::
+        p_i(t + \epsilon/2) &= p_i(t) - (\epsilon/2) d U(q_i(t))/dq_i\\
+        q_i(t + \epsilon) &= q_i(t) +
+                    \epsilon M^{-1}(p_i(t + \epsilon/2)) p_i(t + \epsilon/2)\\
+        p_i(t + \epsilon) &= p_i(t + \epsilon/2) -
+                                (\epsilon/2) d U(q_i(t + \epsilon))/dq_i
+
+    where relativistic mass (a scalar) is,
+
+    .. math::
+        M(p) = m (\Sigma_{i=1}^{d} p_i^2 / (mc^2) + 1)^{0.5}
+
+    In particular, the algorithm we implement follows eqs. in section 2.1 of
+    [1]_.
+
+    Extends :class:`SingleChainMCMC`.
+
+    References
+    ----------
+    .. [1] "Relativistic Monte Carlo". Xiaoyu Lu, Valerio Perrone,
+            Leonard Hasenclever, Yee Whye Teh, Sebastian J. Vollmer,
+            2017, Proceedings of Machine Learning Research.
+    """
+    def __init__(self, x0, sigma0=None):
+        super(RelativisticMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._running = False
+        self._ready_for_tell = False
+
+        # Current point in the Markov chain
+        self._current = None            # Aka current_q in the chapter
+        self._current_energy = None     # Aka U(current_q) = -log_pdf
+        self._current_gradient = None
+        self._current_momentum = None   # Aka current_p
+
+        # Current point in the leapfrog iterations
+        self._momentum = None       # Aka p in the chapter
+        self._position = None       # Aka q in the chapter
+        self._gradient = None       # Aka grad_U(q) in the chapter
+
+        # Iterations, acceptance monitoring, and leapfrog iterations
+        self._mcmc_iteration = 0
+        self._mcmc_acceptance = 0
+        self._frog_iteration = 0
+
+        # Default number of leapfrog iterations
+        self._n_frog_iterations = 20
+
+        # Default integration step size for leapfrog algorithm
+        self._epsilon = 0.1
+        self._step_size = None
+        self._mass = 1
+        self._c = 10
```

Comment:
Don't let @chonlei see this

Reply:
Haha, I'm a physicist too and quite enjoy it.